src/base/http/requestparser.cpp

```diff
@@ -30,314 +31,307 @@
 #include "requestparser.h"
 
 #include <QDebug>
-#include <QDir>
+#include <QRegularExpression>
 #include <QStringList>
 #include <QUrl>
 #include <QUrlQuery>
 
-const QByteArray EOL("\r\n");
-const QByteArray EOH("\r\n\r\n");
+#include "base/utils/string.h"
 
-inline QString unquoted(const QString &str)
+using namespace Http;
+using QStringPair = QPair<QString, QString>;
+
+namespace
 {
-    if ((str[0] == '\"') && (str[str.length() - 1] == '\"'))
-        return str.mid(1, str.length() - 2);
+    const QByteArray EOH = QByteArray(CRLF).repeated(2);
 
-    return str;
-}
+    QByteArray chopIfEndsWith(QByteArray in, const QByteArray &str)
+    {
+        if (in.endsWith(str))
+            in.chop(str.size());
+        return in;
+    }
 
-using namespace Http;
+    QList<QByteArray> split(const QByteArray &in, const QByteArray &sep, const QString::SplitBehavior behavior = QString::KeepEmptyParts)
+    {
+        // mimic QString::split(sep, behavior)
 
-RequestParser::ErrorCode RequestParser::parse(const QByteArray &data, Request &request, uint maxContentLength)
-{
-    return RequestParser(maxContentLength).parseHttpRequest(data, request);
+        if (sep.isEmpty())
+            return {};
+
+        QList<QByteArray> ret;
+
+        int head = 0;
+        while(head < in.size()) {
+            int end = in.indexOf(sep, head);
+            if (end < 0)
+                end = in.size();
+
+            // omit empty parts
+            const QByteArray part((in.constData() + head), (end - head));
+            if (!part.isEmpty() || (behavior == QString::KeepEmptyParts))
+                ret += part;
+
+            head = end + sep.size();
+        }
+
+        return ret;
+    }
 }
 
-RequestParser::RequestParser(uint maxContentLength)
-    : m_maxContentLength(maxContentLength)
+RequestParser::RequestParser()
+    : m_request()
 {
 }
 
-RequestParser::ErrorCode RequestParser::parseHttpRequest(const QByteArray &data, Request &request)
+RequestParser::ParseResult RequestParser::parse(const QByteArray &data)
 {
-    m_request = Request();
+    // Warning! Header names are converted to lowercase
+    return RequestParser().doParse(data);
+}
 
-    // Parse HTTP request header
+RequestParser::ParseResult RequestParser::doParse(const QByteArray &data)
+{
+    // for simplicity, we don't handle malformed requests which use double `LF` as delimiter
     const int headerEnd = data.indexOf(EOH);
     if (headerEnd < 0) {
         qDebug() << Q_FUNC_INFO << "incomplete request";
-        return IncompleteRequest;
+        return {ParseStatus::Incomplete, Request(), 0};
     }
 
-    if (!parseHttpHeader(data.left(headerEnd))) {
+    const QString httpHeaders = QString::fromLatin1(data.left(headerEnd));
```

Comment:
Can we omit unnecessary data copying here?

Reply:
I haven't found a way.

Reply:
Using `QString::fromLatin1(const char *str, int size = -1)` overload.

Reply:
ok, thanks!